File: Python/206_Reverse_Linked_List.py
###iteratively reverse a linked list

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        if head == None or head.next == None:
            return head
        
        prev = None
        while head:
            tmp = head.next
            head.next = prev
            prev = head
            head = tmp
        head = prev
        return head



#recursive method
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        return self.rev_(head)
    
    def rev_(self,node,prev=None):
        if not node:
            return prev
        
        n = node.next
        node.next = prev
        return self.rev_(n,node)

File: C++/206_Reverse_Linked_List.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode *prev = NULL;
    while(head)
        {
          ListNode *tmp;
          tmp = head->next;
          head->next = prev;
          prev = head;
          head = tmp;
        }
        return prev;
    }
};


//recursive method
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
      if(!head || !head->next)
          return head;

      ListNode* node = reverseList(head->next);
      head->next->next = head;
      head->next = NULL;
      return node;


    }
};
